shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -343,4 +345,280 @@ package object ast {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        val maxLenCheck = digitLength <= maxLengthConstant
+
+        if (result < multmin && maxLenCheck && !eFlag)
+          return None
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            if (result < limit + digit && maxLenCheck)
+              return None
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (result < limit)
+      None
+    else {
+      if (negative)
+        Some(result.toLong)
+      else
+        Some(-result.toLong)
+    }
+  }
+
+  private[ast] def toBigInt(value: String): Option[BigInt] = {
+    var decimalFlag = false
+    var result: BigInt = 0
+    var negative = false
+    var char: Char = 0
+    var i = 0
+    var eFlag = false
+    var trailingZeroes: BigInt = 0
+    var negativeEFlag: Boolean = false
+    var resultNegativeEFlag: BigInt = 0
+    var digitLength: BigInt = 0
+    if (value.charAt(0) == '-') {
+      negative = true
+      i += 1
+    }
+
+    while (i < value.length) {
+      char = value.charAt(i)
+      if (char == '.')
+        decimalFlag = true
+      else if ((char | 0x20) == 'e') {
+        eFlag = true
+        val charNext = value.charAt(i + 1)
+        if (charNext == '-')
+          negativeEFlag = true
+        if (negativeEFlag || charNext == '+') {
+          i += 1
+        }
+      } else {
+        if (!(eFlag || decimalFlag))
+          digitLength += 1
+
+        val digit = Character.digit(char, radix)
+
+        if (digit == 0)
+          if (!decimalFlag)
+            trailingZeroes += 1
+          else if (!decimalFlag)
+            trailingZeroes = 0
+
+        if (digit != 0 && (decimalFlag || eFlag))
+          if (!negativeEFlag)
+            return None
+          else {
+            if (trailingZeroes != 0) {
+              resultNegativeEFlag *= radix
+              resultNegativeEFlag += digit
+              if (trailingZeroes >= resultNegativeEFlag) {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (negative)
+      Some(result)
+    else
+      Some(-result)
+  }
+
+  private[ast] def toDouble(value: String): Option[Double] = {
+    try {
+      val asDouble = value.toDouble
+      if (BigDecimal(value, MathContext.UNLIMITED) == BigDecimal(asDouble, MathContext.UNLIMITED))
```

Comment:
I think we are mixing concerns here, the point of `.toDouble` is to get a valid double out of a `JNumber` without losing any data (like any of the other `.to` functions), its not designed to be used for comparisons. For example, I am not seeing how

```scala
BigDecimal("0.142857142857142849") == BigDecimal("0.142857142857142849".toDouble)
```

Is relevant, the issue here is that testing for equality with `Double`'s is problematic and you always need to have a range when testing for quality with a `Double`.

The point of `toDouble` specifically on a `JNumber` is saying is that "if we lose data due to precision or otherwise, it will return `None`". Nothing is preventing a user from going `value.toDouble` if they don't care about losing data, just as nothing is preventing a user from going `value.toInt` (or any other variant). Honestly, if a user is aware of the issues with using `Double` then they are going to know what `JNumber("0.142857142857142849").value.toDouble` is going to do and the risk (i.e. risk of losing accuracy)

Equality of `Double`'s is an orthogonal concern here, and honestly in this area no language gets it really right when comparing `Double`s for equality. In regards to best practice, honestly the only valid best practice I have heard is "don't use doubles unless you really need perfomance" (and this is why clojure and other lisps has ratio type in standard library by default, because the amount of problems typical users trip into when dealing with doubles was enough to justify it)